### PR TITLE
add support for libpressio as an operator

### PR DESCRIPTION
**Pull request by robertu94**
Hi,

I developed [libpressio](https://github.com/CODARcode/libpressio), a generic abstraction for dense tensors for several of the error bounded lossy compressors including SZ, ZFP, and others not currently supported by ADIOS.  While I am aware that there is support for SZ/ZFP already in ADIOS2, this support adds a number of features that currently aren't offered by the current bindings:

+ GPU support for ZFP, and soon for SZ (GPU support for SZ when merged will not require changes to this code).
+ Several currently unsupported options for both SZ and ZFP including those for SZ's "protect_value_range" and new "modes" like ROIBIN-SZ, ZFP's expert mode, and more.
+ Supports additional several additional compressors like SZAuto, fpzip, other image formats via image magic, as well as meta-compressors such as FRaZ/LibPressio-Opt which automatically configure compressors.

I'm relatively sure that I missed some things that I need to tweak to make this play nicely in the ADIOS-2 ecosystem (unit-test jump out to me), so please let me know what I'm missing and what I need to fill in.

I did write an "executable" test based on the ZFP ones which looks like this:

```cpp
  int size, rank;
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  adios2::ADIOS adios(MPI_COMM_WORLD);
  adios2::IO io = adios.DeclareIO("TestIO");

  const int nx=5, ny=5;
  const adios2::Dims 
      shape{nx, ny*static_cast<size_t>(size)},
      start{0, ny*static_cast<size_t>(rank)},
      count{nx, ny};
  size_t totals = size*nx*ny;

  auto temp = io.DefineVariable<float>("temp", shape, start, count, adios2::ConstantDims);
  auto prec = io.DefineVariable<float>("prec", shape, start, count, adios2::ConstantDims);
  auto time_var = io.DefineVariable<double>("time", {1}, {0}, {1}, adios2::ConstantDims);

  adios2::Operator lp_op = adios.DefineOperator("LibPressioOperator", adios2::ops::LossyLIBPRESSIO);
  prec.AddOperation(lp_op, {{"compressor_id", "sz"},
                            {"config:sz:rel_err_bound", "1e-6"},
                            {"config:sz:error_bound_mode_str", "rel"}});
  temp.AddOperation(lp_op, {{"compressor_id", "sz"},
                            {"config:sz:abs_err_bound", "1e-4"},
                            {"config:sz:error_bound_mode_str", "abs"}});
  (void)time_var;

  adios2::Engine engine = io.Open("test.f32", adios2::Mode::Write);

  std::vector<float> t(totals), p(totals);

  if(rank == 0 && size == 0) {
    // ostream.write<int32_t>("size", size);
  }
  for(std::size_t step = 0; step < 4; ++step) {
    auto begin = std::chrono::high_resolution_clock::now();
    std::fill(std::begin(t), std::end(t), static_cast<float>(step));
    std::fill(std::begin(p), std::end(p), static_cast<float>(step));
    auto end = std::chrono::high_resolution_clock::now();
    auto time = static_cast<std::chrono::duration<double>>(end - begin).count();

    engine.BeginStep();
    engine.Put<double>("time", &time);
    engine.Put<float>("temp", t.data());
    engine.Put<float>("prec", p.data());
    engine.EndStep();
  }
  engine.Close();

  //ostream.close();
  MPI_Finalize();

```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -43,6 +43,8 @@ endif()
 # Let windows builds auto-export dll symbols
 set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (ax3l) [inline CMakeLists.txt:46]: This should probably not be part of the PR :)

File: source/adios2/operator/compress/CompressLibPressio.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * it->first ==ompanying file Copyright.txt for details.
```
Comment (ax3l) [inline source/adios2/operator/compress/CompressLibPressio.cpp:3]: Just a typo: looks like a little code paste landed here.

File: source/adios2/toolkit/format/bp/bpOperation/compress/BPLIBPRESSIO.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * accompanying file Copyright.txt for details.
+ *
+ * BPLIBPRESSIO.cpp :
+ *
+ *  Created on: Jul 17, 2018
+ *      Author: William F Godoy godoywf@ornl.gov
```
Comment (ax3l) [inline source/adios2/toolkit/format/bp/bpOperation/compress/BPLIBPRESSIO.cpp:8]: Probably forgotten to update the author :)

File: CMakeLists.txt (hunk 1 above)
Comment (robertu94) [reply CMakeLists.txt]: Good point.

**Comment by JasonRuonanWang** [comment]
Anyone still following this PR? I am very interested to see how this would help reduce the complexity of the compression operators in ADIOS2, as they are getting harder and harder to use and maintain.

**Comment by robertu94** [comment]
I am, but I don't have an active point of contact within Adios to work this from the ORNL side. 

This is my first time offering a contribution to the project, and I will need a little help getting this PR to your standards. I've done some basic testing on my side, but am unsure what else needs to be done.

**Comment by JasonRuonanWang** [comment]
@robertu94 I can help you. We should probably aim for passing the tests first. Can I have write access of your branch so I can push fixes to this PR?

**Comment by robertu94** [comment]
@JasonRuonanWang allow edits by maintainers is checked. Do you need different access?

**Comment by JasonRuonanWang** [comment]
@robertu94 It looks like something weird happened. I was trying to rebase your branch on the current master but somehow it closed this PR. I guess it's probably because this branch had the same branch name as the master branch. Can you make a different branch and submit a PR again? Or you can force-push your local branch and re-open this PR. I will see what I can do with this.

**Comment by eisenhauer** [comment]
You know, I was looking at a rebase briefly last night and ran into the branch name problem as well. @robertu94 or @JasonRuonanWang, you can likely cherry-pick these commits into a new branch with a different name and resubmit the PR.

**Comment by JasonRuonanWang** [comment]
> You know, I was looking at a rebase briefly last night and ran into the branch name problem as well. @robertu94 or @JasonRuonanWang, you can likely cherry-pick these commits into a new branch with a different name and resubmit the PR.

Yeah. I am doing it now :)

**Comment by JasonRuonanWang** [comment]
Moved this to #2742
